Port AccessControlList from Hyrax

no-reply opened the pull request.

Commit:
port AccessControlList from Hyrax

adds ChangeSet based AccessControlList behavior and the grant/revoke DSL.

this is mostly a straight port from Hyrax, with the exception that it doesn't
introduce a dependency on Hyrax::Publisher/Hyrax.publisher. instead, the save
method yields `self` immediately after save, to allow callers to hook in
post-save behavior like publishing. e.g. for Hyrax:

```ruby
class Hyrax::AccessControlList < Hyrax::Acl::AccessControlList
  def save
    super do |acl|
      Hyrax.publisher.publish('object.acl.updated', acl: acl, result: :success)
    end
  end
end
```

Commit:
add tests and documentation for yielding behavior

lsitu commented:
@no-reply Nice! Thanks. 
@marrus-sh @VivianChu Do we want to merge it now?

marrus-sh commented on lib/hyrax/acl/access_control_list.rb:
```diff
@@ -58,12 +104,122 @@ def has_grant?(mode:, agent:)
+      #
+      # @return [Boolean]
+      def pending_changes?
+        change_set.changed?
```
Can we add a test to see if adding an already‐existing permission causes this to be `true`? My guess is that it will, and I’m wondering if we can’t make the grant methods a little more sophisticated and avoid adding to the change set if the permission already exists.

If adding an already‐existing permission doesn’t activate `change_set.changed?` then even better and no changes are needed :P .

no-reply replied:
i'll workup some tests showing this. i think this is the reason `Set` is used, but i can demonstrate that in the test suite before we move forward.

marrus-sh commented on lib/hyrax/acl/access_control_list.rb:
```diff
@@ -15,17 +11,67 @@ module Acl
+      # @param persister [#save] defaults to the configured Hyrax persister
+      # @param query_service [#find_inverse_references_by] defaults to the
+      #   configured Hyrax query service
+      def initialize(resource:, persister:, query_service:)
```
I’m wondering what your feelings are regarding this 🆚 e.g…

```ruby
AccessControlList.from_resource(resource, persister: persister, query_service: query_service)
```

and leaving the initializer taking a `access_control`. Is there a reason we need to keep track of the resource in addition to its access controls?

no-reply replied:
`access_to` depends on `@resource`, or at least the `id`. i think a rewrite could be appropriate. but i do want to call out that a change to the public API on the initializer will definitely block adoption in Hyrax pending a major version of the engine.

marrus-sh commented on lib/hyrax/acl/access_control_list.rb:
```diff
@@ -15,17 +11,67 @@ module Acl
+      #
+      # @return [Boolean]
+      def <<(permission)
+        permission.access_to = resource.id
```
`access_to` gets set on `<<` but not on `delete`… this feels like a bit of an odd discrepancy to me. I’m also not entirely sure what the advantage of these methods are over the DSL ones—what use‐case are they trying to solve?

(I understand that the DSL depends on them now, but it feels like one could just as easily provide ModeEditor with the change set directly and do the implementation there.)

no-reply replied:
i'm uncertain off the top of my head (and couldn't quickly discover with project-wide grep) whether Hyrax has existing code that depends on this. the fundamental use case is: i have user input that results in domain model objects `Hyrax::Permission` and i'd like to set them directly on an object. `acl = Hyrax::Acl::AccessControlList(resource); acl << permissions; acl.save`

the view in development was that the Grant/Revoke DSL was a nicety provided over this lower level service. my experience has been that closing off direct access to low level APIs leads to users reaching through into private ones. since this API is fairly simple, and felt like a prerequisite to implementation for Grant/Revoke, i left it public.

no-reply replied:
the difference between `#<<` and `#delete` is arguably logical:

i never want to add a `Permission` with another target object, so `#<<` casts (maybe it should error instead? i think i felt at the time that adding with an empty target was likely to be the only situation to actually arise, and that the intention in this case would uniformly be to add the permission with the target for this ACL. this is probably not in use, and i think we could safely reconsider.

on the flip side, i don't want `acl << public_read_my_resource ; acl.delete(public_read_other_resource)` to result in an empty permission set.

marrus-sh commented on lib/hyrax/acl/change_set.rb:
```diff
@@ -0,0 +1,51 @@
+    # @example
+    #   Hyrax::ChangeSet(Monograph)
+    def self.ChangeSet(resource_class)
+      klass = (resource_class.to_s + "ChangeSet").safe_constantize || Hyrax::Acl::ChangeSet
```
It feels a little bit weird to be adding these complex change set behaviours to what is otherwise a very minimal gem but I understand if you feel they are necessary. (Are they? The `ChangeSet` used by `AccessControlList` isn’t ever exposed, so is there a problem with it being more minimal? Or is subclassing `AccessControlList` something we’re worried about?)

no-reply replied:
it seemed like it would be more work to undo this than to simply port it in as-is.

i think it would be fine to provide static change sets each for Permission and AccessControl (or even to skip Permission, since we don't use it internally), but i'd rather not block feature parity on a rewrite of this part of the codebase.

marrus-sh commented on spec/hyrax/acl/access_control_list_spec.rb:
```diff
@@ -38,4 +63,139 @@
+  describe 'grant DSL' do
+    let(:mode) { :read }
+    let(:user) { Hyrax::Acl::Agent.new('user1') }
+    let(:group) { Hyrax::Acl::Agent.new("group/public") }
```
There is a `Hyrax::Acl::Group` class now; might be better to use that in this test.

marrus-sh reviewed (changes requested):
Requesting changes because I’d like tests for `#pending_changes?` to clarify its behaviour, but this looks good on the whole.

Commit:
add tests for AccessControlList#pending_changes?

this exercise shows that using `Set` for the property on the `AccessControl`
resource doesn't result in a `Set` on the `ChangeSet`, so adding the same
property twice results in `pending_changes? => true` even though saving won't
result in changes.

is there a fix for this at the `ChangeSet` layer?

Commit:
refactor AccessControlList specs to use `Acl::Group`

Commit:
replace ChangeSet with a static implementation with type behavior

the autogenerated ChangeSet wasn't sensitive to Set type behavior. using a
ChangeSet tailored more precisely for the purpose gives better results on
`#changed?`.

no-reply commented:
@marrus-sh i think this is ready for your re-review

lsitu commented on lib/hyrax/acl/access_control_list.rb:
```diff
@@ -58,12 +104,122 @@ def has_grant?(mode:, agent:)
+        # This defaults to the `:agent_key`, but if that method doesn’t exist,
+        # `:user_key` will be used as a fallback.
+        def id_for(agent:)
+          key = agent.try(:agent_key) || agent.user_key
```
Do we need to add `user_key` attribute/method to Hyrax::Acl::Agent? It only has a single property `name`  and method `agent_key`.

no-reply replied:
@lsitu `user_key` is for compatibility with `Hydra::User`/devise.

it's there for the case when `Hyrax::Acl::Agent` isn't supported by the app.

marrus-sh commented on spec/hyrax/acl/access_control_spec.rb:
```diff
@@ -9,7 +9,7 @@
 
   let(:controlled_resource_class) { Class.new(Valkyrie::Resource) }
-  let(:controlled_resource) { controlled_resource_class.new(id: 'etaoin') }
+ let(:controlled_resource) { controlled_resource_class.new(id: 'etaoin') }
```
Was this intentional?

marrus-sh commented:
@no-reply I think I’d still rather `#<<` throw an error when provided a non‐`nil` `access_to` which doesn’t match the resource id, but otherwise this looks good to me.
